x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Repeat.java

```diff
@@ -0,0 +1,134 @@
+import org.elasticsearch.xpack.esql.EsqlClientException;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.function.OptionalArgument;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.core.type.DataTypes;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.common.unit.ByteSizeUnit.MB;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+
+public class Repeat extends EsqlScalarFunction implements OptionalArgument {
+
+    static final long MAX_REPEATED_LENGTH = MB.toBytes(1);
+
+    private final Expression str;
+    private final Expression number;
+
+    @FunctionInfo(
+        returnType = "keyword",
+        description = "Returns a string constructed by concatenating `string` with itself the specified `number` of times.",
+        examples = @Example(file = "string", tag = "repeat")
+    )
+    public Repeat(
+        Source source,
+        @Param(name = "string", type = { "keyword", "text" }, description = "String expression.") Expression str,
+        @Param(name = "number", type = { "integer" }, description = "Number times to repeat.") Expression number
+    ) {
+        super(source, Arrays.asList(str, number));
+        this.str = str;
+        this.number = number;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isString(str, sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return isType(number, dt -> dt == DataTypes.INTEGER, sourceText(), SECOND, "integer");
+    }
+
+    @Override
+    public boolean foldable() {
+        return str.foldable() && number.foldable();
+    }
+
+    @Evaluator(extraName = "Constant", warnExceptions = { IllegalArgumentException.class })
+    static BytesRef processConstantNumber(BytesRef str, @Fixed int number) {
+        return processInner(str, number);
+    }
```

Comment:
If you are interested in further optimizations you could pass a `@Fixed(includeInToString=false) BytesRef target` that is pre-sized. Then you never have to allocate inside this process. You always know it's big enough.

It's not required. Just having the function is nice. But some folks find this kind of thing fun.

Reply:
I went ahead and add this, definitely seems like a better way to do it!